keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,652 @@
+
+This specification defines a new concept in HTTP, "Alternative Services", that allows an origin server to nominate additional means of interacting with it on the network.
+
+All conformant clusters are required to publish a list of endpoints with the apiserver addresses.In the default implementation, this list of endpoints is generated by a reconcile loop that guarantees that only ready apiserver addresses are present, see https://github.com/kubernetes/kubernetes/tree/master/pkg/controlplane/reconcilers.
+
+The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, this requires that the users of this feature belong to the cluster network. TODO define a way of filtering the headers in the answer, per example, by source IP, certificate name, ... 
+
+
+```sh
+1103 12:30:59.066469 1558329 round_trippers.go:454] GET https://127.0.0.1:44267/api/v1/namespaces/default/pods?limit=500 200 OK in 1 milliseconds
+I1103 12:30:59.066484 1558329 round_trippers.go:460] Response Headers:
+I1103 12:30:59.066491 1558329 round_trippers.go:463]     Cache-Control: no-cache, private
+I1103 12:30:59.066502 1558329 round_trippers.go:463]     Alt-Svc: h2="10.0.0.2:6443", h2="10.0.0.3:6443", h2="10.0.0.4:6443
+```
+
+### Client side HA
+
+Client-go creates a layer on top of the golang net/http library abstracting the communications against the apiservers using [Requests](https://github.com/kubernetes/client-go/blob/master/rest/request.go).
+
+The golang net/http library allows to modify a request in different points:
+- RoundTripper/Transport
+- Dialer
+- DNS dialer
+
+The client-go base-URL is immutable after creation, that guarantees that Requests will not be able to modify it externally.
+
+Client-go already implements some [custom RoundTrippers](https://github.com/kubernetes/client-go/blob/master/transport/round_trippers.go) for some functionalities, like debugging, authentication, impersonation, ...
+
+The proposal is to implement a new round tripper that has a local cache with the list of available apiservers.
+This list can be created manually via configuration (comma-separated list on the Config object) or automatically via the Alt-Svc headers published.
+
+The round tripper behavior, if there are alternative services, will be:
+
+1. Prefer local (same system as the client) alternative services.
+2. Prefer the one in the original request if is listed as an alternative service.
+3. If one alternative service is chosen and succeed, stick to it, only change on network failures.
+4. If the connection against an alternative service fails, block that alternative service for a specified timeout so it will not be retried, retry against the other available alternative services, if everything fails fall back to the original host.
+
+It retries only in case of network failures, that guarantees the idempotency of the request.
```

Comment:
Just L2/3/4 network failures or HTTP failures?  E.g. if I get a 404 or 502 - does client-go try another server?

Reply:
> Just L2/3/4 network failures

this roundtripper only cares about L2/L3/L4, think of it a s dumb L4 proxy.

The http logic is handled by the stdlib and the client-go `Request` methods